addons/web_tour/static/src/tour_service/tour_service.js

```diff
@@ -292,6 +292,24 @@ export const tourService = {
             observer.observe(macroEngine.target, { childList: true, subtree: true });
         }
 
+        /**
+         * Register shadow roots that must be observed by the tour to
+         * the macro engine.
+         *
+         * @param {Tour} tour
+         */
+        function setupShadowObservers(tour) {
+            const shadowDOMs = tour.steps.reduce((acc, step) => {
+                if (step.shadow_dom) {
+                    acc.add(step.shadow_dom);
+                }
+                return acc;
+            }, new Set());
```

Comment:
```suggestion
            const shadowDOMs = new Set(steps.map((s) => s.shadow_dom).filter(Boolean));
```
I know you're just moving this code but it's a good opportunity to clean up. My general advice would be: avoid reduce for anything that's not a sum/product. If you want a filterMap just use filter and map, and if you want something more complicated just create an object and use a regular for loop, it's always much clearer.

Reply:
Agreed, I updated this part it looks much simpler. Thank you for your review!